sdk/js-cloud-server/src/cloudClient.ts

```diff
@@ -0,0 +1,233 @@
+} from './types'
+import { DVCVariable } from './models/variable'
+import { checkParamDefined } from './utils/paramUtils'
+import { dvcDefaultLogger } from './utils/logger'
+import {
+    DVCPopulatedUser,
+    DevCyclePlatformDetails,
+} from './models/populatedUser'
+import {
+    DVCLogger,
+    getVariableTypeFromValue,
+    VariableTypeAlias,
+} from '@devcycle/types'
+import {
+    getAllFeatures,
+    getAllVariables,
+    getVariable,
+    postTrack,
+    ResponseError,
+} from './request'
+import { DevCycleUser } from './models/user'
+
+const castIncomingUser = (user: DevCycleUser) => {
+    if (!(user instanceof DevCycleUser)) {
+        return new DevCycleUser(user)
+    }
+    return user
+}
+
+/**
+ * Handle response errors from API
+ * Suppress 5xx errors and log them instead while returning defaults to the caller
+ * Throw 4XX errors back to the caller to notify of invalid SDK usage
+ * Special handling of 404 error, which indicates "variable not found"
+ * @param err
+ */
+const throwIfUserError = (err: unknown) => {
+    if (err instanceof ResponseError) {
+        const code = err.status
+
+        // throw the error if it indicates there was user error in this call
+        // (e.g. invalid auth credentials or bad user data)
+        if (code !== 404 && code < 500) {
+            throw new Error(
+                `DevCycle request failed with status ${code}. ${
+                    err.message || ''
+                }`,
+            )
+        }
+
+        // Catch non-4xx errors so we can log them instead
+        return
+    }
+
+    if (err instanceof SyntaxError) {
+        // JSON parse error, log instead of throwing
+        return
+    }
+
+    // if not a ResponseError, throw it
+    throw err
+}
+
+export class DevCycleCloudClient {
+    private sdkKey: string
+    private logger: DVCLogger
+    private options: DevCycleOptions
+    private platformDetails: DevCyclePlatformDetails
+
+    constructor(
+        sdkKey: string,
+        options: DevCycleOptions,
+        platformDetails: DevCyclePlatformDetails,
+    ) {
+        this.sdkKey = sdkKey
+        this.logger =
+            options.logger || dvcDefaultLogger({ level: options.logLevel })
+        this.options = options
+        this.platformDetails = platformDetails
+        this.logger.info('Running DevCycle NodeJS SDK in Cloud Bucketing mode')
+    }
+
+    async variable<T extends DVCVariableValue>(
+        user: DevCycleUser,
+        key: string,
+        defaultValue: T,
+    ): Promise<DVCVariable<T>> {
+        const incomingUser = castIncomingUser(user)
+        const populatedUser = DVCPopulatedUser.fromDVCUser(
+            incomingUser,
+            this.platformDetails,
+        )
+
+        checkParamDefined('key', key)
+        checkParamDefined('defaultValue', defaultValue)
+        const type = getVariableTypeFromValue(
+            defaultValue,
+            key,
+            this.logger,
+            true,
+        )
+
+        try {
```

Comment:
Also converted all these functions to async functions from using Promises and `.then()` / `.catch()`